processing/src/main/java/org/apache/druid/query/filter/InDimFilter.java

```diff
@@ -132,31 +145,40 @@ public FilterTuning getFilterTuning()
   @Override
   public byte[] getCacheKey()
   {
-    boolean hasNull = false;
-    for (String value : values) {
-      if (value == null) {
-        hasNull = true;
-        break;
+    if (cacheKey == null) {
+      final List<String> sortedValues = new ArrayList<>(values);
+      sortedValues.sort(Comparator.nullsFirst(Ordering.natural()));
+      final Hasher hasher = Hashing.sha256().newHasher();
```

Comment:
Why did you choose a sha256 hasher? Is it fast to construct a new hasher? or to perform a sha256 hash?

I wrote a test here which fails to produce the same cacheKey even though the values are sorted

```
@Test
  public void testCacheKey()
  {
    final InDimFilter dimFilter1 = new InDimFilter("dim", ImmutableList.of("v1", "v2", "v3"), null);
    final InDimFilter dimFilter2 = new InDimFilter("dim", ImmutableList.of("v3", "v2", "v1"), null);
    Assert.assertEquals(dimFilter1.getCacheKey(), dimFilter2.getCacheKey());
  }
```

Reply:
Sorry about the false alarm. I just realized it's failing because we're comparing 2 byte arrays, but they are in fact equivalent

```
  @Test
  public void testCacheKey()
  {
    final InDimFilter dimFilter1 = new InDimFilter("dim", ImmutableList.of("v1", "v2", "v3"), null);
    final InDimFilter dimFilter2 = new InDimFilter("dim", ImmutableList.of("v3", "v2", "v1"), null);
    byte[] cacheKey1 = dimFilter1.getCacheKey();
    byte[] cacheKey2 = dimFilter2.getCacheKey();
    Assert.assertEquals(Arrays.toString(cacheKey1), Arrays.toString(cacheKey2));
  }
```

Reply:
I believe your test will pass if you use `Assert.assertArrayEquals()` (`getCacheKey()` returns a byte array).

> Why did you choose a sha256 hasher? Is it fast to construct a new hasher? or to perform a sha256 hash?

I chose sha256 since it is long enough to distinguish different `values` sets even though I haven't really tested. I will do some tests.

Reply:
Oh sorry, didn't see your last comment. 👍

Reply:
> I chose sha256 since it is long enough to distinguish different `values` sets even though I haven't really tested. I will do some tests.

👍

I think a simpler solution would be to revert back to the previous way of building a cacheKey, but use `.appendStringsIgnoringOrder()` instead of `.appendStrings()` This way the sort is done once for the list and we don't add any new memory overhead from the current implementation.

Maybe consider creating a new utility function to the CacheKeyBuilder that sorts the list and writes whether or not any nulls existed in the list. This way we can do one pass of the list to build the cacheKey, but even at 10k entries in the list, this might be a micro optimization.

Reply:
The issue with storing all `values` in the cacheKey is that the cache key can become huge as the size of `values` grows which can result in pollution in the cache. 

Regarding using SHA-256, I believe academical research is more trustable than my testing. [According to the probability table, the required number of hashed elements such that probability of at least one hash collision >= 10^−18 is 4.8×10^29 under assumption that the hash function is perfect](https://en.wikipedia.org/wiki/Birthday_problem). This is negligible when it comes to practice even though the SHA-256 hash function is not proven to be perfect.